src/bit_machine/frame.rs

```diff
@@ -15,227 +15,283 @@
-            *p & (1 << (7 - self.abs_pos % 8)) != 0
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u32(&mut self, value: u32, data: &mut [u8]) {
+        for idx in 0..32 {
+            self.write_bit(value & (1 << (31 - idx)) != 0, data);
         }
     }
 
-    pub(crate) fn read_bit(&mut self) -> bool {
-        unsafe {
-            let p = self.data.offset(self.abs_pos / 8);
-            let ret = *p & (1 << (7 - self.abs_pos % 8)) != 0;
-            self.abs_pos += 1;
-            ret
+    /// Write a big-endian u8 value and advance the cursor.
+    pub(crate) fn write_u64(&mut self, value: u64, data: &mut [u8]) {
+        for idx in 0..64 {
+            self.write_bit(value & (1 << (63 - idx)) != 0, data);
         }
     }
 
-    READ_UNSIGNED!(read_u8, u8);
-    READ_UNSIGNED!(read_u16, u16);
-    READ_UNSIGNED!(read_u32, u32);
-    READ_UNSIGNED!(read_u64, u64);
+    /// Move the cursor forward by the given length.
+    pub(crate) fn move_cursor_forward(&mut self, len: usize) {
+        self.cursor += len;
+    }
 
-    /// Write a big-endian u64 value to the current write frame
-    pub(in crate::bit_machine) fn write_u64(&mut self, data: u64) {
-        for idx in 0..64 {
-            self.write_bit(data & (1 << (63 - idx)) != 0);
-        }
+    /// Move the cursor backward by the given length.
+    pub(crate) fn move_cursor_backward(&mut self, len: usize) {
+        self.cursor -= len;
     }
 
-    /// Write a big-endian u32 value to the current write frame
-    pub(in crate::bit_machine) fn write_u32(&mut self, data: u32) {
-        for idx in 0..32 {
-            self.write_bit(data & (1 << (31 - idx)) != 0);
+    /// Copy a bit string of given length from another frame into the present one.
+    pub(crate) fn copy_from(&mut self, other: &Self, len: usize, data: &mut [u8]) {
+        for i in 0..len {
+            let (other_byte_index, other_bit_index) = get_indices(other.cursor + i);
+            let bit = data[other_byte_index] & (1 << (7 - other_bit_index)) != 0;
+            self.write_bit(bit, data);
         }
     }
 
-    /// Write a big-endian u16 value to the current write frame
-    pub(in crate::bit_machine) fn write_u16(&mut self, data: u16) {
-        for idx in 0..16 {
-            self.write_bit(data & (1 << (15 - idx)) != 0);
+    /// Adds
+    pub fn with_data<'a>(&self, data: &'a [u8]) -> FrameData<'a> {
+        FrameData::new(self, data)
+    }
+
+    fn read_unsigned<T>(&mut self, data: &[u8]) -> T
+    where
+        T: From<u8> + Shl<usize, Output = T> + Add<Output = T>,
+    {
+        let (mut self_byte_index, self_bit_index) = get_indices(self.cursor);
+        let mut read_number;
+
+        let number_bits = size_of::<T>() * 8;
+        let number_leading_bits = 8 - self_bit_index;
+        let number_full_bytes = (number_bits - number_leading_bits) / 8;
+        let number_trailing_bits = number_bits - number_leading_bits - number_full_bytes * 8;
```

Comment:
In 372e8a313303e67ab42dc2fe3879850225ffe52b:

nit: it would be a bit more idiomatic to just write `(number_bits - number_leading_bits) % 8` rather than subtracting. (It is a common idiom to define something by `X / Y` then immediately following do `X % Y').

This is a very cool use of Rust generics. I should be more open to trying to use the numeric generics, which I tend to assume are useless because of the lack or sane mathematical guarantees (e.g. the stdlib implements `Add` non-commutatively on some types) and the inability to obtain a generic zero or one.